src/mlpack/methods/lsh_model/lshmodel_impl.hpp

```diff
@@ -0,0 +1,505 @@
+  // generated or estimated in steps 3 and 5.
+  // Gamma distribution for pairwise distances.
+  arma::vec logMeanVec(k + 1), meanLogVec(k + 1), meanVec(k + 1);
+  // Statistics were computed in Train()
+  meanVec(0) = this->meanDist;
+  logMeanVec(0) = this->logMeanDist;
+  meanLogVec(0) = this->meanLogDist;
+  // Train gamma and put in gammaDists[0].
+
+  Timer::Start("fitting_distributions");
+  for (size_t i = 1; i <= k; ++i)
+  {
+    // Use the arithmetic and geometric mean predictors that were trained in
+    // Train() to estimate the statistics for the given datasetSize and k.
+    meanVec(i) = aMeanPredictor.Predict(datasetSize, k);
+    logMeanVec(i) = std::log(meanVec(i));
+    // log(geometricMean) = \frac{1}{n} \sum(lnx_i) = mean(lnx) = meanLog
+    meanLogVec(i) = std::log(gMeanPredictor.Predict(datasetSize, k));
+  }
+  // Fit the distribution.
+  distancesDistribution.Train(logMeanVec, meanLogVec, meanVec);
+  Timer::Stop("fitting_distributions");
+
+  // See if works
+  //GenerateTemplateSequence(3, 0.5, 8);
+
+  // Step 7. Run Binary search on parameter space to minimize selectivity while
+  // keeping recall above minimum.
+
+}
+
+// Fit two predictors, one for arithmetic mean E and one for geometric mean G.
+template <typename SortPolicy, typename ObjectiveFunction>
+void LSHModel<SortPolicy, ObjectiveFunction>::
+ApproximateKNNStatistics(const arma::Col<size_t>& referenceSizes,
+                         const arma::Col<size_t>& kValues,
+                         const arma::mat& Ek,
+                         const arma::mat& Gk)
+{
+  double aError = aMeanPredictor.Train(referenceSizes, kValues, Ek);
+  Log::Info << "L_BFGS Converged for arithmetic mean with error "
+    << aError << "." << std::endl;
+  double gError = gMeanPredictor.Train(referenceSizes, kValues, Gk);
+  Log::Info << "L_BFGS Converged for geometric mean with error "
+    << gError << "." << std::endl;
+}
+
+// Construct and return an LSH object.
+template <typename SortPolicy, typename ObjectiveFunction>
+LSHSearch<SortPolicy> LSHModel<SortPolicy, ObjectiveFunction>::
+LSHObject(const size_t numProjIn,
+          const size_t numTablesIn,
+          const double hashWidthIn,
+          const size_t secondHashSize,
+          const size_t bucketSize)
+{
+  // Values for the object to be created with (specified by user or default).
+  size_t numProjOut = numProjIn;
+  size_t numTablesOut = numTablesIn;
+  double hashWidthOut = hashWidthIn;
+
+  // If not specified by user, set these to the ones we trained for.
+  if (numProjIn == 0)
+    numProjOut = this->numProj;
+
+  if (numTablesIn == 0)
+    numTablesOut = this->numTables;
+
+  if (hashWidthOut == 0.0)
+    hashWidthOut = this->hashWidth;
+
+  LSHSearch<> lsh(*referenceSet, numProjOut, numTablesOut, hashWidthOut,
+      secondHashSize, bucketSize);
+
+  lshObjectVector.push_back(lsh);
+
+  return lshObjectVector[lshObjectVector.size() - 1];
```

Comment:
Why are you preserving the LSH objects you create?  It seems to me that they aren't used anywhere else.